lib/spack/spack/deptypes.py

```diff
@@ -0,0 +1,134 @@
+# New dependency types should be appended.
+LINK = 0b0001
+RUN = 0b0010
+BUILD = 0b0100
+TEST = 0b1000
+
+#: The types of dependency relationships that Spack understands.
+all_types: CanonicalDepTypes = ("build", "link", "run", "test")
+
+#: An iterator of all dependency types
+flag_iterator: Tuple[DepFlag, DepFlag, DepFlag, DepFlag] = (BUILD, LINK, RUN, TEST)
+
+#: A flag with all dependency types set
+all_flag = BUILD | LINK | RUN | TEST
+
+#: Default dependency type if none is specified
+default_deptype: CanonicalDepTypes = ("build", "link")
+
+#: Default dependency type if none is specified
+default_deptype_flag: DepFlag = BUILD | LINK
+
+
+def flag_from_string(s: str) -> DepFlag:
+    if s == "build":
+        return BUILD
+    elif s == "link":
+        return LINK
+    elif s == "run":
+        return RUN
+    elif s == "test":
+        return TEST
+    else:
+        raise ValueError(f"Invalid dependency type: {s}")
+
+
+def type_to_flag(deptype: CanonicalDepTypes) -> DepFlag:
+    """Take a canonical deptype tuple and return a DepFlag."""
+    flag = 0
+    for deptype_str in deptype:
+        flag |= flag_from_string(deptype_str)
+    return flag
+
+
+def flag_to_types(x: DepFlag) -> CanonicalDepTypes:
+    deptype: List[DepType] = []
+    if x & BUILD:
+        deptype.append("build")
+    if x & LINK:
+        deptype.append("link")
+    if x & RUN:
+        deptype.append("run")
+    if x & TEST:
+        deptype.append("test")
+    return tuple(deptype)
+
+
+def flag_to_type(x: DepFlag) -> DepType:
+    if x == BUILD:
+        return "build"
+    elif x == LINK:
+        return "link"
+    elif x == RUN:
+        return "run"
+    elif x == TEST:
+        return "test"
+    else:
+        raise ValueError(f"Invalid dependency type flag: {x}")
+
+
+def canonical_deptype(deptype: DepTypes) -> CanonicalDepTypes:
+    """Convert deptype to a canonical sorted tuple, or raise ValueError.
+
+    Args:
+        deptype: string representing dependency type, or a list/tuple of such strings.
+            Can also be the builtin function ``all`` or the string 'all', which result in
+            a tuple of all dependency types known to Spack.
+    """
+    if deptype in ("all", all):
+        return all_types
+
+    elif isinstance(deptype, str):
+        if deptype not in all_types:
+            raise ValueError(f"Invalid dependency type: {deptype!r}")
+        return (deptype,)  # type: ignore
+
+    elif isinstance(deptype, (tuple, list, set)):
+        deptypes_set = sorted(set(deptype))
+        if any(d not in all_types for d in deptypes_set):
+            raise ValueError(
+                "Invalid dependency types: %s"
+                % ",".join(repr(t) for t in deptypes_set if t not in all_types)
+            )
+        return tuple(deptypes_set)  # type: ignore
+
+    raise ValueError(f"Invalid dependency type: {deptype!r}")
+
+
+def flag_to_chars(depflag: DepFlag) -> str:
```

Comment:
I don't have a really good name, but:
```suggestion
def display_str(depflag: DepFlag) -> str:
```
maybe conveys better the difference with other conversions?

Reply:
I'll just leave this, i think it's fine